frontend/talentsearch/src/js/components/experienceForm/ExperienceForm.tsx

```diff
@@ -0,0 +1,256 @@
+import React from "react";
+import { useIntl } from "react-intl";
+import { toast } from "react-toastify";
+import { SubmitHandler } from "react-hook-form";
+import isEqual from "lodash/isEqual";
+import { BasicForm, TextArea } from "@common/components/form";
+import { commonMessages } from "@common/messages";
+import { getLocale } from "@common/helpers/localize";
+import { navigate } from "@common/helpers/router";
+import useLocalStorage from "@common/hooks/useLocalStorage";
+
+import ProfileFormWrapper from "../applicantProfile/ProfileFormWrapper";
+import ProfileFormFooter from "../applicantProfile/ProfileFormFooter";
+
+import AwardDetailsForm from "../awardDetailsForm/AwardDetailsForm";
+import CommunityExperienceForm from "../communityExperienceForm/CommunityExperienceForm";
+import EducationExperienceForm from "../educationExperienceForm/EducationExperienceForm";
+import PersonalExperienceForm from "../personalExperienceForm/PersonalExperienceForm";
+import WorkExperienceForm from "../workExperienceForm/WorkExperienceForm";
+
+import ExperienceSkills from "./ExperienceSkills";
+import WatchFormValues from "./WatchFormValues";
+import type { Skill } from "../../api/generated";
+import {
+  useGetMyExperiencesQuery,
+  useGetMeQuery,
+  useGetSkillsQuery,
+} from "../../api/generated";
+import applicantProfileRoutes from "../../applicantProfileRoutes";
+
+import type {
+  ExperienceType,
+  AllFormValues,
+  FormValues,
+  ExperienceDetailsSubmissionData,
+  ExperienceMutationResponse,
+  ExperienceQueryData,
+  ExperienceDetailsDefaultValues,
+} from "./types";
+
+import queryResultToDefaultValues from "./defaultValues";
+import formValuesToSubmitData from "./submissionData";
+import useExperienceMutations from "./mutations";
+
+export interface ExperienceFormProps {
+  experienceType: ExperienceType;
+  experience?: ExperienceQueryData;
+  skills: Skill[];
+  onUpdateExperience: (values: ExperienceDetailsSubmissionData) => void;
+}
+
+export const ExperienceForm: React.FunctionComponent<ExperienceFormProps> = ({
+  experience,
+  experienceType,
+  onUpdateExperience,
+  skills,
+}) => {
+  const intl = useIntl();
+  const defaultValues = experience
+    ? queryResultToDefaultValues(experienceType, experience)
+    : { skills: undefined };
+  const [locallySavedForm, setLocallySavedForm] = useLocalStorage<
+    ExperienceDetailsDefaultValues | undefined
+  >(
+    "ts-createExperience", // unique storage key
+    defaultValues, // start form off empty
+  );
```

Comment:
Yeah, if you start one form and navigate to another without submitting, it carries values over

Reply:
> Some thoughts about this form caching... 🤔
> 
> It looks liked the cached values will override the defaultValues as long as the key exists in local storage. I can't test it right now, but it sounds like if I work on one Experience, and then I try to edit a second Experience, I'll be stuck looking at the previous one.
> 
> One way around this would be to use more storage keys. Maybe `` key = `ts-createExperience-${experience ? experience.id : experienceType}`; `` Then editing each existing experience creates its own cache, and Create New has its own for each different type.
> 
> This would create a lot of keys over time, and local storage hangs around forever. Maybe we should use [session storage](https://developer.mozilla.org/en-US/docs/Web/API/Window/sessionStorage) instead for this. That would protect your data if you navigated away accidentally, but not if you closed the tab. But it also stops old data from accruing.
> 
> Another thought - maybe we clear the cache (at this key) when the form is submitted?

A lot of good ideas in this. So right now we are *kind of* clearing it when the form is submitted, however our hook does not allow to actually delete the item, just set it to a null value. We could look into exporting another function from the hook that allows for [removing an item](https://developer.mozilla.org/en-US/docs/Web/API/Storage/removeItem) as well?

As far a session storage goes, is it *immediately* cleared when a tab is closed? I have a bad habit of closing tabs on accident all the time and assume I'm not the only one. If it deletes immediately I would recommend sticking with local storage and just delete the item once the form is submitted.

Reply:
Just thought of a different problem with using unique keys per experience. What if I edit an experience on one computer, which saves it in localStorage. Then I edit it again on another device. The first device will show the old cached data! My first thought is this would be another good reason to use session storage.

Reply:
Okay, so I'll make a `useSessionStorage` hook and update the form to use that instead.

Reply:
Spun off into issue https://github.com/GCTC-NTGC/gc-digital-talent/issues/2530

Reply:
> Spun off into issue #2530

Alright, so are we good to merge this then?

Reply:
Yep @esizer, I don't this discussion to block this PR

Reply:
> As far a session storage goes, is it immediately cleared when a tab is closed? I have a bad habit of closing tabs on accident all the time and assume I'm not the only one. If it deletes immediately I would recommend sticking with local storage and just delete the item once the form is submitted.

Since we're using our own hook, we could use localStorage but build in some logic which includes an expiration time. eg if the hook loads data which is more than an hour old it discards it. That might get around most of these concerns.

Reply:
If we wanted to be REALLY clever we could start including `update_at` timestamps in our api! Maybe that's going too far lol.

Reply:
Sounds good, I like both ideas. Maybe we should add those into the new issue if you haven't already. 

I'll work on resolving the merge conflict then merge shortly.